src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,1168 @@
+      return charmap[rand() % charmapLength];
+    });
+    return result;
+}
+
+std::string FolderMetadata::decryptMetadataKeys(const std::string& encryptedMetadata) const
+{
+    qCInfo(lcCse()) << "Starting to decrypt the metadata key";
+    unsigned char *out = nullptr;
+    size_t outlen = 0;
+    int err = -1;
+
+    auto path = privateKeyPath(_account);
+    auto pathC = qPrintable(path);
+    auto pkeyFile = fopen(pathC, "r");
+    auto key = PEM_read_PrivateKey(pkeyFile, NULL, NULL, NULL);
+    if (!key) {
+        qCInfo(lcCse()) << "Error reading private key";
+    }
+
+    // Data is base64 encoded.
+    qCInfo(lcCse()) << "encryptedMetadata" << encryptedMetadata;
+    auto raw = QByteArray(encryptedMetadata.c_str(), encryptedMetadata.length());
+    auto b64d = QByteArray::fromBase64(raw);
+    auto in = (unsigned char *) b64d.constData();
+    size_t inlen = b64d.length();
+
+    qCInfo(lcCse()) << "Encrypted metadata length: " << inlen;
+
+    /* NB: assumes key in, inlen are already set up
+    * and that key is an RSA private key
+    */
+    auto ctx = EVP_PKEY_CTX_new(key, nullptr);
+    if (!ctx) {
+        qCInfo(lcCse()) << "Could not create the PKEY context.";
+        exit(1);
+    }
+
+    err = EVP_PKEY_decrypt_init(ctx);
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Could not init the decryption of the metadata";
+        exit(1);
+    }
+
+    err = EVP_PKEY_CTX_set_rsa_padding(ctx, 1); //TODO: Make this a class variable.
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Could not set the RSA padding";
+        exit(1);
+    }
+
+    err = EVP_PKEY_decrypt(ctx, NULL, &outlen, in, inlen);
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Could not determine the buffer length";
+        exit(1);
+    } else {
+        qCInfo(lcCse()) << "Size of output is: " << outlen;
+    }
+
+    out = (unsigned char *) OPENSSL_malloc(outlen);
+    if (!out) {
+        qCInfo(lcCse()) << "Could not alloc space for the decrypted metadata";
+        exit(1);
+    }
+
+    err = EVP_PKEY_decrypt(ctx, out, &outlen, in, inlen);
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Could not decrypt the metadata";
+        exit(1);
+    }
+
+    qCInfo(lcCse()) << "Metadata decrypted successfully";
+    const auto ret = std::string((char*) out, outlen);
+
+    return ret;
+}
+
+// AES/GCM/NoPadding (128 bit key size)
+std::string FolderMetadata::encryptJsonObject(const nlohmann::json& obj,const std::string& pass) const {
+    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+    if (!ctx) {
+        qCInfo(lcCse()) << "Coult not create encryption context, aborting.";
+        exit(1);
+    }
+
+    if(!EVP_EncryptInit_ex(ctx, EVP_aes_128_gcm(), NULL, NULL, NULL)) {
+        qCInfo(lcCse()) << "Error initializing encryption aborting.";
+        exit(1);
+    }
+
+    EVP_CIPHER_CTX_set_padding(ctx, 0);
+
+
+
+    unsigned char *iv = (unsigned char *)"0123456789012345";
+    if(!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, 16, NULL)) {
```

Comment:
Once we parse it, we should also set the real length of the IV. It will most likely always be 16. But easier to just get it right once we know it to avoid nasty bugs.